Sources/SwiftShell/Process.swift

```
@@ -0,0 +1,97 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+extension Process {
+
+    // MARK: Public
+
+    @discardableResult
+    /// Executes a shell command.
+    /// - Parameters:
+    ///   - command: The shell command to execute.
+    ///   - shell: The path to the shell from which to execute. Defaults to zsh.
+    ///   - directory: The directory within which the command will execute.
+    public static func execute(
+        _ command: String,
+        from shell: Shell = .zsh,
+        within directory: Directory = .pwd)
+    throws
+    -> String
+    {
+        let script = makeCommand(command, executeWithin: directory)
+        let task = Process()
+        task.launchPath = shell.rawValue
+        task.arguments = ["-c", script]
+
+        let stdout = Pipe()
+        task.standardOutput = stdout
+        let stderr = Pipe()
+        task.standardError = stderr
+
+        try task.run()
+        task.waitUntilExit()
+        guard task.terminationStatus == 0 else {
+            throw ShellError(
+                code: task.terminationStatus,
+                stdout: try stdout.readOutput(),
+                stderr: try stderr.readOutput(),
+                command: command)
+        }
+
+        return try stdout.readOutput()
+    }
+
+    // MARK: Private
+
+    private static func makeCommand(
+        _ command: String,
+        executeWithin directory: Directory)
+    -> String
+    {
+        switch directory {
+        case .pwd:
+            // We're already in the current directory – there's nothing to wrap.
+            return command
+        case .path(let path):
+            return """
+            pushd \(path)
+            \(command)
+            popd
+            """
+        }
+    }
+}
+
+extension Pipe {
+    fileprivate func readOutput() throws -> String {
```

Would it be better to not obfuscate the errors from this method and instead pass them up?

---

isn't that what `throws` is doing? I'm not sure how else we'd want to surface these errors.

---

Yup, I should've been more clear, I just meant the `fileHandleForReading.close()` error is what we're obfuscating here.

---

ohhhh yeah the only reason I'm not throwing that one is I wanted to put the clause into `defer`, and you can't `try` from there.

And given that the docs say you don't need to close the thing anyways... seems like capturing errors closing isn't a super high priority. We can always change our mind later though on this one.